### ISSUE TITLE
Does parcel use dotenv, or dotenv flow? The documentation is incomplete and confusing

### ISSUE THREAD

**Issue by rsimp** [parcel-bundler/website#1013]
Is Parcel using [dotenv](https://github.com/motdotla/dotenv) as it claims, or [dotenv-flow](https://github.com/kerimdzhanov/dotenv-flow)? It links to `dotenv` but supports multiple env files in the same manner as `dotenv-flow`, which is not included behavior with `dotenv` and actively warns against it. However `dotenv-flow` has specific rules on which env files have higher precedence. Command line env variables have the highest precedence, and `.env.local` is still loaded for test. This is the behavior that next.js and create react app uses.

I think parcel needs to go with one or the other, link to it, and use the same logic. If not remove the link and just document it's own rules : What files can be source controlled? What files overwrite others? Are command line env variables overwritten if defined etc.

Also what is the rationale behind not using `.env.local` for test? Env variables loaded via jenkins/aws/etc are higher priority than ones defined in env files with `dotenv-flow`. `.env.local` will have unsourced secrets and may need to be different than the actual test environment (dev/qa/staging etc), but API keys may still be necessary for any integration or end-to-end testing. Precedence order should make sure it all works correctly which is the main concern when using multiple `.env` files. I'd probably end up with two local env files with the same configuration and I'm not sure I'm understanding what the benefit is.

**Comment by rsimp** [comment]
Also `dotenv-flow` supports using an unsourced `.env` file and having a sourced `.env.defaults`. Is this behavior supported as well? Do you see my point? Either you need to document the full behavior, or document every case where the behavior deviates. It might be better to just say you took inspiration from `dotenv-flow`, but documentation is lacking either way.

Right now the behavior mimics `dotenv-flow` more than `dotenv` but doesn't quite work the same, and is missing a link to the documentation for `dotenv-flow` which answers questions the Environment variables section doesn't answer.

**Comment by mischnic** [comment]
This is where the loading happens: https://github.com/parcel-bundler/parcel/blob/e294eafd9a49c056fb4b223c5ace5c0653428ede/packages/core/core/src/loadDotEnv.js#L10

**Comment by rsimp** [comment]
Right, so basically both respect command line variables. You're using `dotenv` and `dotenv-expand` but you're not following the `dotenv` philosophy of only having one `.env` file. You're using multiple env files tied to `NODE_ENV` with a precedence order, like with `dotenv-flow`, but `.env.local` is treated differently for `NODE_ENV=test`. `.env.defaults` is not supported.

The precedence order is implied but not explicitly stated in the docs, especially that command line variables have highest precedence. `dotenv` should document that it doesn't overwrite variables but also only accepts a single env file so probably worth explicitly stating everything. The documentation also mentions nothing about source control.

TBH I'm not sure why you guys don't just use `dotenv-flow` instead, less code, more features, easy to link to. I also still don't understand disabling `.env.local` on test but at least it's well documented.

I still think it's a bit confusing that you've sort of hand rolled your own `dotenv-flow` with slight differences. I would look at their docs as a good guideline for [precedence order](https://github.com/kerimdzhanov/dotenv-flow#variables-overwritingpriority) and [source control](https://github.com/kerimdzhanov/dotenv-flow#files-under-version-control). Precedence order is almost documented just need a blurb about increasing precedence listed in your table, and that existing command line variables have the highest priority. A section on source control would be extremely helpful as `dotenv` simply tells you not to source anything.

**Comment by rsimp** [comment]
Hmm, so I see that the code is actually the same as in [create-react-app](https://github.com/facebook/create-react-app/blob/fd8c5f7b1b1d19d10d24cc2f9fdfc110585dc030/packages/react-scripts/config/env.js#L28) and [next.js](https://github.com/vercel/next.js/blob/6611bfd31cf236342e6d31a33cf3ce50c44b2e59/packages/next-env/index.ts#L84). Same behavior for test with the same comment verbatim: 

> Don't include `.env.local` for `test` environment since normally you expect tests to produce the sameresults for everyone

 I thought they used to use `dotenv-flow` but I guess they switched at some point. 

The behavior is definitely more in alignment with `dotenv-flow` and disregards advice on [multiple files](https://github.com/motdotla/dotenv#should-i-have-multiple-env-files) and [source control](https://github.com/motdotla/dotenv#should-i-commit-my-env-file).

I think the biggest thing missing here is a section on [source control](https://github.com/kerimdzhanov/dotenv-flow#files-under-version-control), and maybe expand on [precedence order](https://github.com/kerimdzhanov/dotenv-flow#variables-overwritingpriority). Both [create-react-app](https://create-react-app.dev/docs/adding-custom-environment-variables/#adding-development-environment-variables-in-env) and [next.js](https://nextjs.org/docs/basic-features/environment-variables#default-environment-variables) have sections in their documentation going over both in more detail.

**Comment by rsimp** [comment]
Even more curious, your precedence order is the the same as `dotenv-flow`: (low to high precedence) `.env` -> `.env.local` -> `.env.[environment]` -> `.env.[environment].local`. Whereas create-react-app and next.js are slightly different: `.env` -> `.env.[environment]` -> `.env.local` -> `.env.[environment].local`. 

If `.env.local` can override `.env.test`, it would make sense that it wouldn't be loaded, and why the difference. Although its curious you have the same precedence order as `dotenv-flow` but don't load `.env.local` for test. 

Reading the create-react-app [code](https://github.com/facebook/create-react-app/blob/fd8c5f7b1b1d19d10d24cc2f9fdfc110585dc030/packages/react-scripts/config/env.js#L25), it seems the different precedence order comes from a ruby dotenv implementation, which I think is what also excluded `.env.local` from test.

**Comment by mrcoles** [comment]
I just ran into this question too. How about this proposed edit to the docs? https://github.com/parcel-bundler/website/pull/1079
